pkg/events/check.go

```diff
@@ -156,7 +160,29 @@ func (ce *CheckEvent) getRepo(ctx context.Context, cloneURL, branchName string)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to clone repo")
 	}
+
 	ce.clonedRepos[reposKey] = repo
+
+	// this is all to ensure that one request for `main` and a follow up request for HEAD
+	// (or the other way around) retrieves the same repo
+	if branchName == "" || branchName == "HEAD" {
+		remoteHeadBranchName, err := repo.GetRemoteHead()
+		if err != nil {
+			return repo, errors.Wrap(err, "failed to determine remote head")
+		}
+
+		repo.BranchName = remoteHeadBranchName
+		ce.clonedRepos[generateRepoKey(cloneURL, remoteHeadBranchName)] = repo
+	} else if _, ok := ce.clonedRepos[generateRepoKey(cloneURL, "HEAD")]; !ok {
```

Comment:
Maybe simplify this a bit? There's some pretty deeply-nested logic here; splitting this out and maybe using early exit might make this more readable, IMHO.